File: Stack/SimplifyPath.go
// LeetCode 71 - Simplify Path
// Time Complexity: O(n) | Space Complexity: O(n)
package main

import "strings"

func simplifyPath(path string) string {
	stack := make([]string, 0)

	for _, part := range strings.Split(path, "/") {
		if part == "" || part == "." {
			continue
		}
		if part == ".." {
			if len(stack) > 0 {
				stack = stack[:len(stack)-1]
			}
		} else {
			stack = append(stack, part)
		}
	}

	return "/" + strings.Join(stack, "/")
}

File: Stack/SimplifyPath.java
// LeetCode 71 - Simplify Path
// Time Complexity: O(n) | Space Complexity: O(n)
import java.util.ArrayDeque;
import java.util.Deque;

public class SimplifyPath {
    public String simplifyPath(String path) {
        Deque<String> stack = new ArrayDeque<>();

        for (String part : path.split("/")) {
            if (part.isEmpty() || ".".equals(part)) {
                continue;
            }
            if ("..".equals(part)) {
                stack.pollLast();
            } else {
                stack.offerLast(part);
            }
        }

        return "/" + String.join("/", stack);
    }
}
